src/test/analyzer_test.ts

```diff
@@ -523,4 +546,211 @@ suite('Analyzer', () => {
+      const waitTimes: number[] = [];
+
+      const reproducableWait = () => new Promise<void>((resolve) => {
+        const waitTime = waitTimes.shift();
+        if (waitTime == null) {
+          throw new Error(
+              'Was asked for more random waits than the ' +
+              'given array of wait times');
+        }
+        setTimeout(resolve, waitTime);
+      });
+      await editorSimulator(reproducableWait);
+    });
+
+    suite('deterministic tests', () => {
+      // Deterministic tests extracted from various failures of the above random
+      // test.
+
+      class Deferred<T> {
+        promise: Promise<T>;
+        resolve: (result: T) => void;
+        reject: (error: Error) => void;
+        constructor() {
+          this.promise = new Promise((resolve, reject) => {
+            this.resolve = resolve;
+            this.reject = reject;
+          });
+        }
+      }
+      class KeyedQueue<Key, Result> {
+        private _requests = new Map<Key, Deferred<Result>[]>();
+        private _results = new Map<Key, Result[]>();
+        async request(key: Key): Promise<Result> {
+          const results = this._results.get(key) || [];
+          if (results.length > 0) {
+            return results.shift()!;
+          }
+          const deferred = new Deferred<Result>();
+          const deferreds = this._requests.get(key) || [];
+          this._requests.set(key, deferreds);
+          deferreds.push(deferred);
+          return deferred.promise;
+        }
+
+        /**
+         * Resolves the next unfulfilled request for the given key with the
+         * given value.
+         */
+        resolve(key: Key, value: Result) {
+          const requests = this._requests.get(key) || [];
+          if (requests.length > 0) {
+            const request = requests.shift()!;
+            request.resolve(value);
+            return;
+          }
+          const results = this._results.get(key) || [];
+          this._results.set(key, results);
+          results.push(value);
+        }
+
+        toString() {
+          return JSON.stringify({
+            openRequests: Array.from(this._requests.keys()),
+            openResponses: Array.from(this._results.keys())
+          });
+        }
+      }
+      class DeterministicUrlLoader implements UrlLoader {
+        outstandingRequests = new KeyedQueue<string, string>();
+        canLoad(_url: string) {
+          return true;
+        }
+
+        async load(url: string) {
+          return this.outstandingRequests.request(url);
+        }
+      }
+      class NoopUrlLoader implements UrlLoader {
+        canLoad() {
+          return true;
+        }
+        async load(): Promise<string> {
+          throw new Error(
+              `Noop Url Loader isn't supposed to be actually called.`);
+        }
+      }
+
+      test('two edits of the same file back to back', async() => {
+        const analyzer = new Analyzer({urlLoader: new NoopUrlLoader()});
```

Comment:
For regression tests on tricky bugs like this, it's nice to have a comment that describes how this was triggering a bug, or what the failure condition was, or maybe a link to an issue that this tests against.